 1187  sbatch --wrap "kallisto quant -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Reads/combined_PE1.fq /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Reads/combined_PE2.fq " -c 4 -N 1 --mem=20GB -o out_kallisto -e run_info_kallisto -J Cele_kallisto
 1197  sbatch --wrap "kallisto quant -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Reads/combined_PE1.fq /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Reads/combined_PE2.fq " -c 4 -N 1 --mem=20GB -o out_kallisto -e run_info_kallisto -J Dmelo_kallisto
 1211  sbatch --wrap "kallisto quant -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Reads/combined_PE1.fq /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Reads/combined_PE2.fq " -c 4 -N 1 --mem=20GB -o out_kallisto -e run_info_kallisto -J Homo_kallisto
 1231  sbatch --wrap "kallisto quant --fr-stranded -t 4 -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Reads/combined_PE1.fq.gz /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Reads/combined_PE2.fq.gz " -c 4 -N 1 --mem=20GB -o out_kallisto -e run_info_kallisto -J AT_kallisto
 1238  sbatch --wrap "kallisto quant --rf-stranded -t 4 -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Reads/combined_PE1.fq.gz /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Athaliana/Reads/combined_PE2.fq.gz " -c 4 -N 1 --mem=20GB -o out_kallisto -e run_info_kallisto -J AT_kallisto
 1250  sbatch --wrap "kallisto quant -t 8 -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Reads/combined_PE1.fq /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Dmelanogaster/Reads/combined_PE2.fq " -c 8 -N 1 --mem=40GB -o out_kallisto -e run_info_kallisto -J Dmelo_kallisto
 1267  history |grep "sbatch"
 1268  sbatch --wrap "kallisto quant -t 8  -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Reads/combined_PE1.fq /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Celegans/Reads/combined_PE2.fq " -c 8 -N 1 --mem=40GB -o out_kallisto -e run_info_kallisto -J Cele_kallisto
 1275  sbatch --wrap "kallisto quant -t 8 -i /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Reference/reference.cdnas.kallisto -o /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Kallisto /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Reads/combined_PE1.fq /usr/users/ga002/caims/group-ga/Projects/Mikado/From_060716/Hsapiens/Reads/combined_PE2.fq " -c 8 -N 1 --mem=40GB -o out_kallisto -e run_info_kallisto -J Homo_kallisto
 1328  sbatch "source star-2.4.1c;STAR   --runThreadN 16   --genomeDir /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/   --readFilesIn /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz      --readFilesCommand zcat      --outSAMtype BAM   Unsorted      --outSAMstrandField intronMotif   --outSAMattributes NH   HI   AS   nM   NM   MD      --outFilterMismatchNmax 4   --outFilterIntronMotifs RemoveNoncanonical   --alignIntronMax 100000   --alignTranscriptsPerReadNmax 1000000   --alignSJoverhangMin 12   --alignSJDBoverhangMin 12   --alignEndsType EndToEnd" -c 16 -N 1 --mem=50GB -J STAR_newsimu -o out_info -e run_info
 1329  sbatch --wrap "source star-2.4.1c;STAR   --runThreadN 16   --genomeDir /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/   --readFilesIn /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz      --readFilesCommand zcat      --outSAMtype BAM   Unsorted      --outSAMstrandField intronMotif   --outSAMattributes NH   HI   AS   nM   NM   MD      --outFilterMismatchNmax 4   --outFilterIntronMotifs RemoveNoncanonical   --alignIntronMax 100000   --alignTranscriptsPerReadNmax 1000000   --alignSJoverhangMin 12   --alignSJDBoverhangMin 12   --alignEndsType EndToEnd" -c 16 -N 1 --mem=50GB -J STAR_newsimu -o out_info -e run_info
 1373  sbatch --wrap "source star-2.4.1c;STAR   --runThreadN 16   --genomeDir /tgac/workarea/group-ga/Projects/Mikado/From_060716/Athaliana/Indices   --readFilesIn /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz      --readFilesCommand zcat      --outSAMtype BAM   Unsorted      --outSAMstrandField intronMotif   --outSAMattributes NH   HI   AS   nM   NM   MD      --outFilterMismatchNmax 4   --outFilterIntronMotifs RemoveNoncanonical   --alignIntronMax 100000   --alignTranscriptsPerReadNmax 1000000   --alignSJoverhangMin 12   --alignSJDBoverhangMin 12   --alignEndsType EndToEnd" -c 16 -N 1 --mem=50GB -J STAR_newsimu -o out_info -e run_info
 1388  sbatch --wrap "source star-2.4.1c;STAR   --runThreadN 16   --genomeDir /tgac/workarea/group-ga/Projects/Mikado/From_060716/Athaliana/Indices   --readFilesIn /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz      --readFilesCommand zcat      --outSAMtype BAM   Unsorted      --outSAMstrandField intronMotif   --outSAMattributes NH   HI   AS   nM   NM   MD      --outFilterMismatchNmax 4   --outFilterIntronMotifs RemoveNoncanonical   --alignIntronMax 100000   --alignTranscriptsPerReadNmax 1000000   --alignSJoverhangMin 12   --alignSJDBoverhangMin 12   --alignEndsType EndToEnd" -c 16 -N 1 --mem=50GB -J STAR_newsimu -o out_info -e run_info
 1399  sbatch --wrap "tophat --library-type fr-secondstrand -I 100000 -r 30 -p 16 --min-anchor-length 12 --max-multihits 20 -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat /tgac/workarea/group-ga/Projects/Mikado/From_060716/Athaliana/Indices/tair10 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz" -c 16 -N 1 --mem=50GB -o out_info -e run_info 
 1406  sbatch --wrap "tophat --library-type fr-secondstrand -I 100000 -r 30 -p 16 --min-anchor-length 12 --max-multihits 20 -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat /tgac/workarea/group-ga/Projects/Mikado/From_060716/Athaliana/Indices/tair10 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz" -c 16 -N 1 --mem=50GB -o out_info -e run_info -J TophatNSIR
 1448  sbatch --wrap "source star-2.4.1c;STAR   --runThreadN 16   --genomeDir /tgac/workarea/group-ga/Projects/Mikado/From_060716/Athaliana/Indices   --readFilesIn /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz      --readFilesCommand zcat      --outSAMtype BAM   Unsorted      --outSAMstrandField intronMotif   --outSAMattributes NH   HI   AS   nM   NM   MD      --outFilterMismatchNmax 4   --outFilterIntronMotifs RemoveNoncanonical   --alignIntronMax 100000   --alignTranscriptsPerReadNmax 1000000   --alignSJoverhangMin 12   --alignSJDBoverhangMin 12   --alignEndsType EndToEnd" -c 16 -N 1 --mem=50GB -J STAR_newsimu -o out_info -e run_info
 1476  sbatch --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASS-NS --mem=10GB 
 1482  sbatch --wrap "source cufflinks-2.1.1;cufflinks -o STAR_Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam" -c 8 -N 1 -o out.info -e run.info -J Cuff-NS --mem=10GB 
 1485   sbatch --wrap "source stringtie-1.03; stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8" c 8 -N 1 -o out.info -e run.info -J STie-NS --mem=10GB
 1486   sbatch --wrap "source stringtie-1.03; stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8" c 8 -N 1 -o out.info -e run.info -J Stie-NS --mem=10GB
 1487   sbatch --wrap "source stringtie-1.03; stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " c 8 -N 1 -o out.info -e run.info -J Stie-NS --mem=10GB
 1488   sbatch --wrap "source stringtie-1.03; stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " c 8 -N 1 -o out.info -e run.info  --mem=10GB
 1489   sbatch --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " c 8 -N 1 -o out.info -e run.info  -J Stie-NS --mem=10GB
 1490   sbatch --wrap "stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " c 8 -N 1 -o out.info -e run.info  -J Stie-NS --mem=10GB
 1491  sbatch --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8" c 8 -N 1 -o out.info -e run.info  -J Stie-NS --mem=10GB
 1492  sbatch --wrap "ls" c 8 -N 1 -o out.info -e run.info  -J Stie-NS --mem=10GB
 1493   sbatch --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J Stie-NS --mem=10GB
 1499  sbatch -wrap "samtools sort /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-alignment.bam " -c 8 -N 1 --mem=20GB
 1505  sbatch --wrap "samtools sort Aligned.out.bam -o sorted-alignment.bam" -c 8 -N 1 --mem=20GB
 1512  sbatch --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/STAR/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB
 1515  sbatch --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/alignment.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/STAR/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TrinityNS
 1520   sbatch --dependency=afterok:1585042 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-alignment.bam.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J Stie-NS --mem=10GB
 1540  sbatch --dependency=afterok:1584930 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSNSIR --mem=10GB 
 1543  sbatch --dependency=afterok:1584930 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/TopHat/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TnityNSIR
 1546  sbatch --dependency=afterok:1584930 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CuffNST --mem=10GB 
 1560  sbatch --wrap "samtools sort Aligned.out.bam -o sorted-alignment" -c 8 -N 1 --mem=20GB
 1578  sbatch  --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/alignment.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSSTNSI --mem=10GB 
 1581  sbatch --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/STAR/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddataIR0.1 -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/alignment.bam" -c 8 -N 1 -o out.info -e run.info -J CuffSTNSI --mem=10GB 
 1584   sbatch --dependency=afterok:1585084 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieSTNSI --mem=10GB
 1588  sbatch --dependency=afterok:1584930 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/alignment.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/STAR/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TnitySTNSI
 1590  sbatch  --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/alignment.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/STAR/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TnitySTNSI
 1613  sbatch --dependency=afterok:1584930 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSTNSI --mem=10GB 
 1617  sbatch --dependency=afterok:1584930 --wrap "samtools sort accepted_hits.bam -o sorted-accepted_hits" -c 8 -N 1 --mem=20GB -J sortTopNSIR
 1623   sbatch --dependency=afterok:1585122 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/sorted-accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNSI --mem=10GB
 1628  sbatch --dependency=afterok:1584930 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddataIR0.1 -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CuffTNSI --mem=30GB 
 1633  sbatch --dependency=afterok:1584930 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/TopHat/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TnityTNSI
 1646  sbatch --dependency=afterok:1584885 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSTNS --mem=10GB 
 1662  sbatch --wrap "samtools sort Aligned.out.bam -o sorted-accepted_hits" -c 8 -N 1 --mem=20GB -J sortSTNS
 1688  sbatch --wrap "samtools sort alignment.bam sorted-alignment" -c 8 -N 1 --mem=20GB -J sortSTNS
 1716  sbatch --dependency=afterok:1584885 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CuffTNS --mem=30GB 
 1719  sbatch --dependency=afterok:1584885 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/TopHat/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TnityTNS
 1725  sbatch --wrap "samtools sort accepted_hits.bam sorted-accepted_hits" -c 8 -N 1 --mem=20GB -J sortTNS
 1727   sbatch --dependency=afterok:1585155 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/sorted-accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1730   sbatch --dependency=afterok:1584885:1585155 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/sorted-accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1738  sbatch --dependency=afterok:1584885 --wrap "samtools sort accepted_hits.bam sorted-accepted_hits" -c 8 -N 1 --mem=20GB -J sortTNS
 1742   sbatch --dependency=afterok:1584885:1585165 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/sorted-accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1749  sbatch --dependency=afterok:1584930 --wrap "samtools sort accepted_hits.bam sorted-accepted_hits" -c 8 -N 1 --mem=20GB -J sortTNS
 1758   sbatch --dependency=afterok:1584930:1585168 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/sorted-accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1765   sbatch --dependency=afterok:1584930 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1778   sbatch --dependency=afterok:1584885 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1781  sbatch --dependency=afterok:1584885 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSTNS --mem=10GB 
 1783  sbatch --dependency=afterok:1584885 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/TopHat/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TnityTNS
 1786  sbatch --dependency=afterok:1584885 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CuffTNS --mem=30GB 
 1807  sbatch --dependency=afterok:1584930 --wrap "samtools sort Aligned.out.bam sorted-Aligned.out" -c 8 -N 1 --mem=20GB -J sortSTNS
 1809  sbatch  --wrap "samtools sort Aligned.out.bam sorted-Aligned.out" -c 8 -N 1 --mem=20GB -J sortSTNS
 1828  sbatch --dependency=afterok:1585227 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-Aligned.out.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/STAR/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TrinitySNS
 1836  sbatch --dependency=afterok:1585227 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Assemblies/STAR/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-Aligned.out.bam " -c 8 -N 1 -o out.info -e run.info -J CuffSNS --mem=30GB 
 1841  sbatch --dependency=afterok:1584885 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-Aligned.out.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSNS --mem=10GB 
 1849  sbatch --dependency=afterok:1585227 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-Aligned.out.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLASSNS --mem=10GB 
 1852   sbatch --dependency=afterok:1585227 --wrap "source stringtie-1.03;stringtie  /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated/Alignment/STAR/sorted-Aligned.out.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StieTNS --mem=10GB
 1880  sbatch --dependency=afterok:1585144 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-Aligned.out.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLNSI --mem=10GB 
 1883   sbatch --dependency=afterok:1585227 --wrap "source stringtie-1.03;stringtie  /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StNSI --mem=10GB
 1886  sbatch --dependency=afterok:1585144 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-alignment.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLNSI --mem=10GB 
 1889   sbatch --dependency=afterok:1585144 --wrap "source stringtie-1.03;stringtie  /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-alignment.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J StNSI --mem=10GB
 1892  sbatch --dependency=afterok:1585144 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/STAR/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata_IR0.1 -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-alignment.bam " -c 8 -N 1 -o out.info -e run.info -J CFNSI --mem=30GB 
 1898  sbatch --dependency=afterok:1585144 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/STAR/sorted-alignment.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/STAR/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TrinitySNS
 1908  sbatch --dependency=afterok:1584930 --wrap "source class-2.12;class_run.py /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam --force -p 8 -c \"-F 0.05\" > Class.gtf" -c 8 -N 1 -o out.info -e run.info -J CLTNSI --mem=10GB
 1911  sbatch --dependency=afterok:1584885 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata_IR0.1 -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CFTNSI --mem=30GB 
 1913  sbatch --dependency=afterok:1584930 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata_IR0.1 -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CFTNSI --mem=30GB 
 1916  sbatch --dependency=afterok:1584930 --wrap "source cufflinks-2.1.1;cufflinks -o /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/TopHat/Cufflinks -p 8 -b /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reference/genome.fa  -I 100000 -u --library-type fr-unstranded  -L NewSimulateddata_IR0.1 -F 0.05 /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam" -c 8 -N 1 -o out.info -e run.info -J CFTNSI --mem=30GB 
 1921  sbatch --dependency=afterok:1584930 --wrap "source stringtie-1.03;stringtie /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam -l Stringtie -f 0.05 -m 200 -o StringTie.gtf -p 8 " -c 8 -N 1 -o out.info -e run.info  -J STNSI --mem=10GB
 1925  sbatch --dependency=afterok:1584930 --wrap "source jre-7u71;source trinityrnaseq-r20140717;Trinity --genome_guided_use_bam /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Alignment/TopHat/accepted_hits.bam --genome_guided_CPU 8 --genome_guided_max_intron 100000  --output /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Assemblies/TopHat/Trinity/ --seqType fq --JM 30G --left /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_1.fastq.gz --right /tgac/workarea/group-ga/Projects/Mikado/From_060716/NewSimulated_IR0.1/Reads/sim_2.fastq.gz " -c 8 -N 1 --mem=40GB -J TTNSI
 1928  history |grep "sbatch"|less
 1929  history |grep "sbatch" > /tgac/workarea/group-ga/Projects/Mikado/From_060716/CLI_All.txt
